composer/loggers/progress_bar_logger.py

```diff
@@ -21,8 +22,20 @@
 _IS_TRAIN_TO_KEYS_TO_LOG = {True: ['loss/train'], False: ['metrics/eval/Accuracy']}
 
 
+def rank_zero_only(fn: Callable) -> Callable:
```

Comment:
I do appreciate the simplicity of a decorator like this, but am concerned about a naive implementation, as it makes distributed errors much harder to debug in code like the following:

```python

def bar():
    dist.barrier()

@rank_zero_only
def foo():
   bar()
```
The above code snippet will hang, but it is unclear and hard to debug why. We had errors before like this, which is why I removed the `RankZeroCallback` class a while back.

That said, I do like the idea of having a rank zero context manager that causes all torch.dist calls to immediately error if used within context manager, and then we can replace all `if dist.get_global_rank() == 0` with a `with dist.global_rank_zero_only():`. I would prefer a context manager over a decorator, as it draws more visual attention to the code that it is being executed only on rank zero, but am also open to a function / class that could be used as both.

As for an implementation, it could look something like this:


```python
# composer/utils/dist.py

_DIST_ENABLED = True

@contextlib.contextmanager
def global_rank_zero_only(enabled: bool = True):
    global _DIST_ENABLED
    was_enabled = _DIST_ENABLED
    _DIST_ENABLED = enabled
    try:
        yield
    finally:
        _DIST_ENABLED = was_enabled

_DIST_DISABLED_ERROR_MSG = "Distributed functions cannot be called when used inside `global_rank_zero_only()`"

# Then, call dist wrappers (except the getters) would need to have something like this:
def barrier():
    global _DIST_ENABLED
    if not _DIST_ENABLED:
        raise RuntimeError(_DIST_DISABLED_ERROR_MSG)
    ...
```

Then, caller code could look like:

```python
def something():
    with global_rank_zero_only():
        print("Hello!")
```

Anyways, I think adding this decorator / context manager and replacing the existing `if dist.get_global_rank() == 0` checks is outside the scope of this PR.